### DISCUSSION TITLE
FetchContent and submodules

### DISCUSSION THREAD

**Discussion by Lord-Kamina** [performous/performous#719] [Ideas]
I know I'm beating a dead horse here, but given:

1) I understand people's reluctance to work with submodules.

2) CMake FetchContent works as intended but it's relatively hacky and obscure.

3) The idea behind moving was to make things easier for everyone.

Wouldn't it be a better solution to find the middle ground by restoring the submodules, but resorting to CMake FetchContent if they haven't been populated properly? I feel like this would make the intention much clearer and would prevent things like what happened with #713

**Comment by Baklap4** [comment]
Don't get me wrong, i love the use of submodules. And i do think they're quite easy to use once you get your head around the concept.
But in the past, people complained about submodules being not so handy / hard to understand..
Then we had an inbetween moment, where stuff got fuzzy
Then we got rid of submodules and just relied on system libraries using our cmake modules structure as the project used for like ever.

And now we're indeed back at the point about how to handle our dependencies.

In terms of dependencies, for the majority of the dependencies we rely on the libraries the OS provides us. Yet for _one_ (aubio) we have given the user the choice: Wanna use system libraries or the brand new one from latest git by compiling it ourselves.

Personally i think it's a good thing to give the developer/user a choice: use the ,most likely, outdated libraries provided by the OS or use self-built libraries which are close to  or are latest.

I agree that we should make a decision what we want, and then update our cmake modules with the choice we made.

If we want to make it ourselves easier we just only rely on the libraries provided by the OS, but we have to manage different versions of libraries as not every platform has the same version running. This is already nicely done with for example libxml++2.6 - 5.0

If we want more consistency and cleaner code we should build every dependency ourselves, which makes the build-procedure way longer/complicated which is not ideal with CI/CD in mind since caching is hard to figure out how to do it right especially with all the dependencies we have: around 500-700MB of dependencies...

**Reply by Lord-Kamina** [reply]
I think our main focus should be to make it as possible for ourselves (but also others) to build the project.
At the same time, I don't want to pollute our actual development folders with dependencies (like would have been the case with adding the json header).

I think searching the system is generally a good default alternative, but we need to have a fallback. That's where either submodules/fetchcontent come in (or combining both).

The reason I'm proposing going back to submodules (backed by fetchcontent so nobody has issues checking them out) is that it makes the whole thing more transparent. Because you can see they're supposed to be there by looking at the tree.

The way we're currently doing it obfuscates what's really going on unless you dive into the CMakeLists files.

**Reply by yoda-jm** [reply]
I generally don't like fetching dependency while compiling for the following reasons:
- this is the job of a packager, to know exactly how to compile what and with what options
- for security reasons, I dont expect a "configure" step to download stuff on internet
- it can break the reproducability of the compilations for packages and required internet access

I have nothing agains submodule (I just find them hard to understand in the tree compared to having some downloads in the build directory.

I would suggest the following:
- keep FetchContent for the moment: it is in the CMake and is properly advertise
- force to use flag ALLOW_DEPENDENCY_DOWNLOAD to allow FetchContent (make it explicit when something unusual is done)
